src/Symfony/Component/Utf8/Utf8Trait.php

```diff
@@ -0,0 +1,354 @@
+
+    public static function fromCodePoint(int ...$codes)
+    {
+        $string = '';
+        foreach ($codes as $code) {
+            if (0x80 > $code %= 0x200000) {
+                $string .= chr($code);
+            } elseif (0x800 > $code) {
+                $string .= chr(0xC0 | $code >> 6).chr(0x80 | $code & 0x3F);
+            } elseif (0x10000 > $code) {
+                $string .= chr(0xE0 | $code >> 12).chr(0x80 | $code >> 6 & 0x3F).chr(0x80 | $code & 0x3F);
+            } else {
+                $string .= chr(0xF0 | $code >> 18).chr(0x80 | $code >> 12 & 0x3F).chr(0x80 | $code >> 6 & 0x3F).chr(0x80 | $code & 0x3F);
+            }
+        }
+
+        $instance = new static();
+        $instance->string = $string;
+
+        return $instance;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function __toString()
+    {
+        return $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isEmpty(): bool
+    {
+        return '' === $this->string;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function explode(string $delimiter, int $limit = null): array
+    {
+        if ('' === $delimiter) {
+            throw new InvalidArgumentException('Passing an empty delimiter is not supported by this method. Use getIterator() method instead.');
+        }
+
+        $chunks = array();
+        foreach (preg_split('/'.preg_quote($delimiter).'/u', $this->string, $limit ?: -1) as $i => $string) {
+            $chunks[$i] = $chunk = clone $this;
+            $chunk->string = $string;
+        }
+
+        return $chunks;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toLowerCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtolower($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toUpperCase(): self
+    {
+        $result = clone $this;
+        $result->string = mb_strtoupper($this->string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function toFoldedCase(bool $full = true)
+    {
+        $string = str_replace(self::$commonCaseFold[0], self::$commonCaseFold[1], $this->string);
+
+        if ($full) {
+            static $fullCaseFold = false;
+            $fullCaseFold or $fullCaseFold = require __DIR__.'/Resources/data/caseFolding_full.php';
+            $string = str_replace($fullCaseFold[0], $fullCaseFold[1], $string);
+        }
+
+        $result = clone $this;
+        $result->string = mb_strtolower($string, 'UTF-8');
+
+        return $result;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function trim(string $charsList = null): self
+    {
+        if (!preg_match('//u', $charsList)) {
+            throw new InvalidArgumentException('Given chars list is not a valid UTF-8 encoded string.');
+        }
+
+        $charsList = $charsList ? preg_quote($charsList) : '[:space:]';
```

Comment:
As I wrote in my general comment, that interface is broken anyways. I cannot pass an en quad (` `) to `Bytes` because it probably breaks there, but I can pass it to this method. This means that I cannot treat any instance of `GenericStringInterface` equally and will always require an `instanceof` check.